lxd/db/openfga/openfga.go

```diff
@@ -28,6 +31,102 @@ func NewOpenFGAStore(clusterDB *db.Cluster) storage.OpenFGADatastore {
+		ReadUsersetTuplesCache:    make(map[string][]*openfgav1.Tuple),
+	}
+}
+
+// ReadReadStartingFromUserCache reads tuples from the cache.
+func (c *OpenFGACache) ReadReadStartingFromUserCache(filter storage.ReadStartingWithUserFilter) (bool, string, []*openfgav1.Tuple, error) {
+	b, err := json.Marshal(filter)
+	if err != nil {
+		return false, "", nil, err
+	}
+
+	tuples, ok := c.ReadStartingFromUserCache[string(b)]
+	if ok {
+		return true, string(b), tuples, nil
+	}
+
+	return false, string(b), nil, nil
+}
+
+// WriteReadStartingFromUserCache writes tuples to the cache.
+func (c *OpenFGACache) WriteReadStartingFromUserCache(cacheKey string, tuples []*openfgav1.Tuple) {
+	c.ReadStartingFromUserCache[cacheKey] = tuples
+}
+
+// JSONReadUsersetTuplesFilter is an intermediate struct for serializing/deserializing the storage.ReadUsersetTuplesFilter.
+// This is needed because we can't deserialize the storage.ReadUsersetTuplesFilter directly: it contains an interface field that can't be deserialized.
+type JSONReadUsersetTuplesFilter struct {
+	Object                      string `json:"object"`
+	Relation                    string `json:"relation"`
+	AllowedUserTypeRestrictions []struct {
+		Type      string  `json:"type"`
+		Relation  *string `json:"relation"`
+		Wildcard  *bool   `json:"wildcard"`
+		Condition string  `json:"condition"`
+	} `json:"allowed_user_type_restrictions"`
+}
+
+// ReadReadUsersetTuplesCache reads tuples from the cache.
+func (c *OpenFGACache) ReadReadUsersetTuplesCache(filter storage.ReadUsersetTuplesFilter) (bool, string, []*openfgav1.Tuple, error) {
+	jsonFilter := JSONReadUsersetTuplesFilter{
+		Object:   filter.Object,
+		Relation: filter.Relation,
+	}
+
+	if len(filter.AllowedUserTypeRestrictions) > 0 {
+		jsonFilter.AllowedUserTypeRestrictions = make([]struct {
+			Type      string  `json:"type"`
+			Relation  *string `json:"relation"`
+			Wildcard  *bool   `json:"wildcard"`
+			Condition string  `json:"condition"`
+		}, len(filter.AllowedUserTypeRestrictions))
+
+		for i, r := range filter.AllowedUserTypeRestrictions {
+			restriction := &jsonFilter.AllowedUserTypeRestrictions[i]
+			restriction.Type = r.Type
+			restriction.Condition = r.Condition
+
+			switch v := r.RelationOrWildcard.(type) {
+			case *openfgav1.RelationReference_Relation:
+				relation := v.Relation
+				restriction.Relation = &relation
+			case *openfgav1.RelationReference_Wildcard:
+				isWildcard := true
+				restriction.Wildcard = &isWildcard
+			}
+		}
+	}
+
+	b, err := json.Marshal(jsonFilter)
+	if err != nil {
+		return false, "", nil, err
+	}
+
+	tuples, ok := c.ReadUsersetTuplesCache[string(b)]
+	if ok {
+		return true, string(b), tuples, nil
+	}
+
+	return false, string(b), nil, nil
+}
+
+// WriteReadUsersetTuplesCache writes tuples to the cache.
+func (c *OpenFGACache) WriteReadUsersetTuplesCache(cacheKey string, tuples []*openfgav1.Tuple) {
+	c.ReadUsersetTuplesCache[cacheKey] = tuples
```

Comment:
Currently we allow only one call to the datastore at a time (see OpenFGA server options in `lxd/auth/drivers/openfga.go`). Given that we now have caching at the datastore level and that some calls to the datastore don't require database calls at all, I can see us removing that restriction to improve performance. Could you please add a `sync.RWMutex` to these caches so that they are thread safe? Thanks.